src/any_agent/serving/server.py

```diff
@@ -27,8 +30,34 @@ def _get_a2a_server(
     return A2AStarletteApplication(agent_card=agent_card, http_handler=request_handler)
 
 
-def serve_a2a(server: A2AStarletteApplication, host: str, port: int) -> None:
+def _create_server(
+    app: A2AStarletteApplication, host: str, port: int, log_level: str = "warning"
+) -> uvicorn.Server:
+    config = uvicorn.Config(app.build(), host=host, port=port, log_level=log_level)
+    return uvicorn.Server(config)
+
+
+async def serve_a2a_async(
+    server: A2AStarletteApplication, host: str, port: int, log_level: str = "warning"
+) -> tuple[asyncio.Task[Any], uvicorn.Server]:
+    """Provide an A2A server to be used in an event loop."""
+    uv_server = _create_server(server, host, port)
+    task = asyncio.create_task(uv_server.serve())
+    while not uv_server.started:  # noqa: ASYNC110
+        await asyncio.sleep(0.1)
+    return (task, uv_server)
+
+
+def serve_a2a(
+    server: A2AStarletteApplication, host: str, port: int, log_level: str = "warning"
+) -> None:
     """Serve the A2A server."""
-    import uvicorn
 
-    uvicorn.run(server.build(), host=host, port=port)
+    # Note that the task should be kept somewhere
+    # because the loop only keeps weak refs to tasks
+    # https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task
+    async def run() -> None:
+        (task, _) = await serve_a2a_async(server, host, port, log_level)
+        await task
+
+    return asyncio.get_event_loop().run_until_complete(run())
```

Comment:
Sorry if I missed something.

Do we really need to do this? We do the wrapping in the `run` / `run_async` because some frameworks didn't provide a sync API.

Couldn't we just use the uvicorn sync API?

Reply:
We can go from an async api to a sync api, but not vice versa. In tests, and in a foreseeable future, we need to set up a server in an async context to allow other async tasks to be run; for example, a client and a server, or more than one server.

A sync run call will just block the process, which works for simple scenarios but prevents more complex setups.

Reply:
If I understand, @daavoo  isn't questioning the need for an async server, we're on the same page that we need serve_async. The question is whether serve() should call serve_async under the hood, or should have it's own uvicorn.run command that uses the sync uvicorn command.

Personally, I am ok with dealing with this question at a later point.